File: typescript/src/swagger20-reference.ts
import {
  isSwagger20Object,
  member,
  pathItemMemberResource,
  resourceBase,
  stringMember,
  type Swagger20LoadOptions,
  type Swagger20Method,
  type Swagger20PathItem,
  type Swagger20ReferenceGraphContract,
  type Swagger20Resolution,
  type Swagger20ResolutionMemo,
  type Swagger20Resource,
} from "./swagger20-model.js";
import { parseSwagger20Resource } from "./swagger20-loader.js";

/** @internal */
export function newSwagger20ResolutionMemo(): Swagger20ResolutionMemo {
  return { active: new Set(), pathActive: new Set(), done: new Map() };
}

/** Lazy JSON Reference draft-03 resource graph for one Swagger 2.0 source. */
export class Swagger20ReferenceGraph implements Swagger20ReferenceGraphContract {
  readonly selfContained: boolean;
  readonly allowExternalRefs: boolean;
  private readonly fetchFn: typeof globalThis.fetch | undefined;
  private readonly signal: AbortSignal | undefined;
  private readonly resources = new Map<string, Swagger20Resource>();
  private readonly pending = new Map<string, Promise<Swagger20Resource>>();

  constructor(options: Swagger20LoadOptions, selfContained: boolean) {
    this.fetchFn = options.fetch;
    this.signal = options.signal;
    this.selfContained = selfContained;
    this.allowExternalRefs = options.allowExternalRefs !== false;
  }

  rebind(options: Swagger20LoadOptions): Swagger20ReferenceGraph {
    return new Swagger20ReferenceGraph(options, this.selfContained);
  }

  rememberResolvedResource(requested: string | undefined, retrieval: string | undefined, root: unknown): Swagger20Resource {
    const resource = { requested, retrieval, root };
    if (requested) this.resources.set(resourceKey(requested), resource);
    if (retrieval) this.resources.set(resourceKey(retrieval), resource);
    if (!requested && !retrieval) this.resources.set("", resource);
    return resource;
  }

  async resolveReference(
    raw: string,
    from: Swagger20Resource,
    memo: Swagger20ResolutionMemo = newSwagger20ResolutionMemo(),
  ): Promise<Swagger20Resolution> {
    let target: URL;
    try {
      const base = resourceBase(from);
      const fragmentOnly = raw.startsWith("#");
      if (base) target = new URL(raw, base);
      else if (fragmentOnly) target = new URL(raw, "swagger20:self");
      else {
        target = new URL(raw);
        if (this.selfContained) {
          throw new Error(`reference ${JSON.stringify(raw)} cannot resolve: embedded Swagger 2.0 content with no co-present location must be self-contained`);
        }
      }
    } catch (error: unknown) {
      if (error instanceof Error && error.message.includes("self-contained")) throw error;
      throw new Error(`Swagger 2.0 reference ${JSON.stringify(raw)} is not a URI`, { cause: error });
    }

    const fragment = target.hash.slice(1);
    target.hash = "";
    const resourceURI = target.protocol === "swagger20:" ? "" : target.href;
    const canonical = `${resourceKey(resourceURI)}#${fragment}`;
    const completed = memo.done.get(canonical);
    if (completed) return completed;
    if (memo.active.has(canonical)) return { resource: from, cycle: true };
    memo.active.add(canonical);
    try {
      const resource = await this.resourceForReference(resourceURI, from);
      let result: Swagger20Resolution = { node: resource.root, resource };
      if (fragment !== "") {
        const decoded = decodeURIComponent(fragment);
        if (!decoded.startsWith("/")) {
          throw new Error(`Swagger 2.0 reference ${JSON.stringify(raw)} has unsupported non-JSON-Pointer fragment`);
        }
        result = await this.resolvePointer(resource, decoded, memo);
      }
      memo.done.set(canonical, result);
      return result;
    } finally {
      memo.active.delete(canonical);
    }
  }

  async resolvePathItem(
    item: Swagger20PathItem,
    method: Swagger20Method,
    memo: Swagger20ResolutionMemo = newSwagger20ResolutionMemo(),
  ): Promise<Swagger20PathItem> {
    const reference = stringMember(item.raw, "$ref");
    if (!reference.present) return item;
    if (!reference.valid || reference.value === "") {
      throw new Error("selected Swagger 2.0 Path Item has an invalid $ref");
    }
    const key = `${resourceKey(resourceBase(item.resource) ?? "")}|${reference.value}|${method}`;
    if (memo.pathActive.has(key)) {
      throw new Error(`selected Swagger 2.0 Path Item reference cycle leaves ${method} unresolved`);
    }
    memo.pathActive.add(key);
    try {
      const resolved = await this.resolveReference(reference.value!, item.resource, memo);
      if (resolved.cycle) {
        throw new Error(`selected Swagger 2.0 Path Item reference cycle leaves ${method} unresolved`);
      }
      if (!isSwagger20Object(resolved.node)) {
        throw new Error("selected Swagger 2.0 Path Item $ref does not resolve to an object");
      }
      const target = await this.resolvePathItem({ raw: resolved.node, resource: resolved.resource }, method, memo);
      if (member(item.raw, method).present && member(target.raw, method).present) {
        throw new Error(`selected Swagger 2.0 Path Item has undefined adjacent/ref collision at ${method}`);
      }
      if (member(item.raw, "parameters").present && member(target.raw, "parameters").present) {
        throw new Error("selected Swagger 2.0 Path Item has undefined adjacent/ref collision at parameters");
      }
      const merged = { ...target.raw };
      const resources = new Map<string, Swagger20Resource>();
      for (const name of Object.keys(target.raw)) resources.set(name, pathItemMemberResource(target, name));
      for (const [name, value] of Object.entries(item.raw)) {
        if (name === "$ref" || Object.hasOwn(merged, name)) continue;
        merged[name] = value;
        resources.set(name, pathItemMemberResource(item, name));
      }
      return { raw: merged, resource: target.resource, memberResources: resources };
    } finally {
      memo.pathActive.delete(key);
    }
  }

  private async resourceForReference(uri: string, from: Swagger20Resource): Promise<Swagger20Resource> {
    const key = resourceKey(uri);
    if (key === "") return from;
    const fromKeys = [from.requested, from.retrieval].filter(Boolean).map((value) => resourceKey(value!));
    if (fromKeys.includes(key)) return from;
    const cached = this.resources.get(key);
    if (cached) return cached;
    if (this.selfContained) {
      throw new Error(`reference ${JSON.stringify(uri)} cannot resolve: embedded Swagger 2.0 content with no co-present location must be self-contained`);
    }
    if (!this.allowExternalRefs) throw new Error(`external Swagger 2.0 reference ${JSON.stringify(uri)} is disabled`);
    const existing = this.pending.get(key);
    if (existing) return existing;
    const load = this.readResource(uri).finally(() => this.pending.delete(key));
    this.pending.set(key, load);
    return load;
  }

  private async readResource(uri: string): Promise<Swagger20Resource> {
    const fetchFn = this.fetchFn ?? globalThis.fetch;
    if (!fetchFn) throw new Error(`no fetch implementation is available for Swagger 2.0 resource ${JSON.stringify(uri)}`);
    const response = await fetchFn(uri, { signal: this.signal });
    if (!response.ok) throw new Error(`load Swagger 2.0 resource ${JSON.stringify(uri)}: HTTP ${response.status}`);
    const root = parseSwagger20Resource(await response.text());
    const retrieval = response.url || uri;
    return this.rememberResolvedResource(uri, retrieval, root);
  }

  private async resolvePointer(
    resource: Swagger20Resource,
    pointer: string,
    memo: Swagger20ResolutionMemo,
  ): Promise<Swagger20Resolution> {
    if (pointer === "") return { node: resource.root, resource };
    if (!pointer.startsWith("/")) throw new Error(`fragment ${JSON.stringify(pointer)} is not an RFC 6901 JSON Pointer`);
    let current: Swagger20Resolution = { node: resource.root, resource };
    for (const encoded of pointer.slice(1).split("/")) {
      if (!wellFormedPointerToken(encoded)) throw new Error(`fragment ${JSON.stringify(pointer)} contains malformed RFC 6901 escape`);
      const token = decodePointerToken(encoded);
      while (true) {
        if (isSwagger20Object(current.node)) {
          if (Object.hasOwn(current.node, token)) {
            current.node = current.node[token];
            break;
          }
          const reference = stringMember(current.node, "$ref");
          if (!reference.valid) throw new Error(`JSON Pointer token ${JSON.stringify(token)} identifies no object member`);
          const resolved = await this.resolveReference(reference.value!, current.resource, memo);
          if (resolved.cycle) return resolved;
          current = resolved;
          continue;
        }
        if (Array.isArray(current.node)) {
          if (!/^(0|[1-9][0-9]*)$/u.test(token)) throw new Error(`JSON Pointer token ${JSON.stringify(token)} identifies no array member`);
          const index = Number(token);
          if (index >= current.node.length) throw new Error(`JSON Pointer token ${JSON.stringify(token)} identifies no array member`);
          current.node = current.node[index];
          break;
        }
        throw new Error(`JSON Pointer token ${JSON.stringify(token)} cannot be applied to a scalar`);
      }
    }
    return current;
  }
}

/** @internal */
export function wellFormedPointerToken(value: string): boolean {
  for (let index = 0; index < value.length; index++) {
    if (value[index] === "~" && value[index + 1] !== "0" && value[index + 1] !== "1") return false;
    if (value[index] === "~") index++;
  }
  return true;
}

/** @internal */
export function decodePointerToken(value: string): string {
  return value.replaceAll("~1", "/").replaceAll("~0", "~");
}

/** @internal */
export function escapePointerToken(value: string): string {
  return value.replaceAll("~", "~0").replaceAll("/", "~1");
}

function resourceKey(value: string): string {
  if (value === "") return "";
  const url = new URL(value);
  url.hash = "";
  return url.href;
}

File: go/internal/runtime/swagger20_reference.go
package openapiclient

import (
	"fmt"
	"net/url"
	"strings"
)

type swagger20Resolution struct {
	node     any
	resource *swagger20Resource
	cycle    bool
}

type swagger20ResolutionMemo struct {
	active     map[string]bool
	pathActive map[string]bool
	done       map[string]swagger20Resolution
}

func newSwagger20ResolutionMemo() *swagger20ResolutionMemo {
	return &swagger20ResolutionMemo{
		active: map[string]bool{}, pathActive: map[string]bool{}, done: map[string]swagger20Resolution{},
	}
}

func (g *swagger20ReferenceGraph) resolveReference(raw string, from *swagger20Resource, memo *swagger20ResolutionMemo) (swagger20Resolution, error) {
	parsed, err := url.Parse(raw)
	if err != nil {
		return swagger20Resolution{}, fmt.Errorf("Swagger 2.0 reference %q is not a URI: %w", raw, err)
	}
	base := from.base()
	fragmentOnly := parsed.Scheme == "" && parsed.Host == "" && parsed.Path == "" && parsed.RawQuery == ""

	var target *url.URL
	switch {
	case fragmentOnly:
		if base != nil {
			target = base.ResolveReference(parsed)
		} else {
			target = &url.URL{Fragment: parsed.Fragment, RawFragment: parsed.RawFragment}
		}
	case base != nil:
		target = base.ResolveReference(parsed)
	case parsed.IsAbs():
		if g.selfContained {
			return swagger20Resolution{}, fmt.Errorf("reference %q cannot resolve: embedded Swagger 2.0 content with no co-present location must be self-contained", raw)
		}
		target = parsed
	default:
		return swagger20Resolution{}, fmt.Errorf("relative reference %q has no absolute Swagger 2.0 artifact base", raw)
	}

	resourceURI := cloneURL(target)
	canonical := artifactResourceKey(resourceURI) + "#" + target.EscapedFragment()
	if result, ok := memo.done[canonical]; ok {
		return result, nil
	}
	if memo.active[canonical] {
		return swagger20Resolution{resource: from, cycle: true}, nil
	}
	memo.active[canonical] = true
	defer delete(memo.active, canonical)

	resource, err := g.resourceForReference(resourceURI, from)
	if err != nil {
		return swagger20Resolution{}, err
	}
	result := swagger20Resolution{node: resource.root, resource: resource}
	if target.Fragment != "" {
		if !strings.HasPrefix(target.Fragment, "/") {
			return swagger20Resolution{}, fmt.Errorf("Swagger 2.0 reference %q has unsupported non-JSON-Pointer fragment", raw)
		}
		result, err = g.resolvePointer(resource, target.Fragment, memo)
		if err != nil {
			return swagger20Resolution{}, fmt.Errorf("resolve Swagger 2.0 reference %q: %w", raw, err)
		}
	}
	memo.done[canonical] = result
	return result, nil
}

func (g *swagger20ReferenceGraph) resourceForReference(uri *url.URL, from *swagger20Resource) (*swagger20Resource, error) {
	key := artifactResourceKey(uri)
	if key == "" {
		return from, nil
	}
	if from != nil && (key == artifactResourceKey(from.requested) || key == artifactResourceKey(from.retrieval)) {
		return from, nil
	}
	g.mu.RLock()
	cached := g.resources[key]
	g.mu.RUnlock()
	if cached != nil {
		return cached, nil
	}
	if g.selfContained {
		return nil, fmt.Errorf("reference %q cannot resolve: embedded Swagger 2.0 content with no co-present location must be self-contained", uri)
	}
	if !g.allowExternalRefs {
		return nil, fmt.Errorf("external Swagger 2.0 reference %q is disabled", uri)
	}
	data, err := g.read(uri)
	if err != nil {
		return nil, err
	}
	root, err := parseSwagger20Resource(data)
	if err != nil {
		return nil, err
	}
	return g.rememberResource(uri, root), nil
}

// resolvePointer implements RFC 6901 over the referenced document. JSON
// Reference draft-03 transclusion is honored when a pointer token runs below
// a Reference Object: the reference target is substituted and traversal
// continues at the same token.
func (g *swagger20ReferenceGraph) resolvePointer(resource *swagger20Resource, pointer string, memo *swagger20ResolutionMemo) (swagger20Resolution, error) {
	if pointer == "" {
		return swagger20Resolution{node: resource.root, resource: resource}, nil
	}
	if !strings.HasPrefix(pointer, "/") {
		return swagger20Resolution{}, fmt.Errorf("fragment %q is not an RFC 6901 JSON Pointer", pointer)
	}
	current := swagger20Resolution{node: resource.root, resource: resource}
	for _, encoded := range strings.Split(pointer[1:], "/") {
		if !wellFormedPointerToken(encoded) {
			return swagger20Resolution{}, fmt.Errorf("fragment %q contains malformed RFC 6901 escape", pointer)
		}
		token := decodePointerToken(encoded)
		for {
			switch typed := current.node.(type) {
			case map[string]any:
				if child, present := typed[token]; present {
					current.node = child
					goto nextToken
				}
				ref, hasRef := typed["$ref"].(string)
				if !hasRef {
					return swagger20Resolution{}, fmt.Errorf("JSON Pointer token %q identifies no object member", token)
				}
				resolved, err := g.resolveReference(ref, current.resource, memo)
				if err != nil {
					return swagger20Resolution{}, err
				}
				if resolved.cycle {
					return resolved, nil
				}
				current = resolved
			case []any:
				index, ok := rawSequenceIndex(token, len(typed))
				if !ok {
					return swagger20Resolution{}, fmt.Errorf("JSON Pointer token %q identifies no array member", token)
				}
				current.node = typed[index]
				goto nextToken
			default:
				return swagger20Resolution{}, fmt.Errorf("JSON Pointer token %q cannot be applied to a scalar", token)
			}
		}
	nextToken:
	}
	return current, nil
}

func (g *swagger20ReferenceGraph) resolvePathItem(item swagger20PathItem, method string, memo *swagger20ResolutionMemo) (swagger20PathItem, error) {
	ref := item.reference()
	if !ref.present {
		return item, nil
	}
	if !ref.valid || ref.value == "" {
		return swagger20PathItem{}, fmt.Errorf("selected Swagger 2.0 Path Item has an invalid $ref")
	}
	key := artifactResourceKey(item.resource.base()) + "|" + ref.value + "|" + method
	if memo.pathActive[key] {
		return swagger20PathItem{}, fmt.Errorf("selected Swagger 2.0 Path Item reference cycle leaves %s unresolved", method)
	}
	memo.pathActive[key] = true
	defer delete(memo.pathActive, key)
	resolved, err := g.resolveReference(ref.value, item.resource, memo)
	if err != nil {
		return swagger20PathItem{}, err
	}
	if resolved.cycle {
		return swagger20PathItem{}, fmt.Errorf("selected Swagger 2.0 Path Item reference cycle leaves %s unresolved", method)
	}
	targetObject, ok := resolved.node.(map[string]any)
	if !ok {
		return swagger20PathItem{}, fmt.Errorf("selected Swagger 2.0 Path Item $ref does not resolve to an object")
	}
	target, err := g.resolvePathItem(swagger20PathItem{raw: swagger20Object(targetObject), resource: resolved.resource}, method, memo)
	if err != nil {
		return swagger20PathItem{}, err
	}

	if _, adjacentMethod := item.raw.member(method); adjacentMethod {
		if _, referencedMethod := target.raw.member(method); referencedMethod {
			return swagger20PathItem{}, fmt.Errorf("selected Swagger 2.0 Path Item has undefined adjacent/ref collision at %s", method)
		}
	}
	if _, adjacentParameters := item.raw.member("parameters"); adjacentParameters {
		if _, referencedParameters := target.raw.member("parameters"); referencedParameters {
			return swagger20PathItem{}, fmt.Errorf("selected Swagger 2.0 Path Item has undefined adjacent/ref collision at parameters")
		}
	}

	merged := make(swagger20Object, len(target.raw)+len(item.raw))
	memberResources := make(map[string]*swagger20Resource, len(target.raw)+len(item.raw))
	for name, value := range target.raw {
		merged[name] = value
		memberResources[name] = target.resourceFor(name)
	}
	for name, value := range item.raw {
		if name == "$ref" {
			continue
		}
		if _, collision := merged[name]; collision {
			// Collisions in methods and fields unused by this selected target are
			// confined there; the referenced declaration remains authoritative
			// for this target's resolved view.
			continue
		}
		merged[name] = value
		memberResources[name] = item.resourceFor(name)
	}
	return swagger20PathItem{raw: merged, resource: target.resource, memberResources: memberResources}, nil
}
